### ISSUE TITLE
internal algorithm

### ISSUE THREAD

**Issue by applecv3** [lior-k/fast-elasticsearch-vector-scoring#58]
Hi, I just want to know which type of KNN (like HNSW, LSH, and so forth) you built in this plugin.

**Comment by lior-k** [comment]
The plug-in uses pure cosine-similarity or dot-product to compare vectors.
So the K nearest neighbors it returns are the exact K, not any assessment
like LSH and others

**Comment by applecv3** [comment]
Thank you for your answer! So.. let me ask you some more.
Do you mean naive KNN searching algorithm by "pure cosine-similarity"?
Is it taking O(N) time complexity? (where N is the number of documents to explore when computing cosine similarity).
If so, I'm not sure how your plugin works faster than the others and I saw you mentioned that "I gained this substantial speed improvement by using the lucene index directly".
Does that imply all the secrets(?) about how this plugin works fast?

**Comment by lior-k** [comment]
Yes, it uses brute force to calculate cosine-similarity. Meaning O(n)
It is not faster than hnswlib or fasis etc...
It is faster then other ES plugins that did the same brute force
calculations. The only difference was using the lucene engine.
You can see the code :-)

BTW - Amazon has an hnswlib implementation on their manages ES
implementation. It should be much faster than this but it has limitations

**Comment by applecv3** [comment]
Thank you so much! I really appreciate it. Have a good day!

**Comment by lior-k** [comment]
BTW, we use k-means with this plug-in inorder to traverse only the input
vector nearest clusters instead of the entire corpus.

**Comment by zpmmehrdad** [comment]
@lior-k Hi,

Whats is the difference between this repo and the native ES vector scoring?
Which one is faster?

Thanks

**Comment by lior-k** [comment]
Never tested. This plugin existes way before the official support.
If you do test the performance differences please let us all know 🙏

**Comment by Shengwuyou** [comment]
Whether the plug-in can perform algorithm configuration, use brute force to calculate cosine similarity, not suitable for high-efficiency scenarios # @lior-k
